### PR TITLE
Add a warning to the end of the chef run for EOL releses

### PR DESCRIPTION
**Pull request by tas50**
We really want to encourage people to upgrade. One way we can do this is
to just warn them that they're on an EOL release. This simple warning
might nudge them to upgrade their client. It works off the base EOL
data of Chef 15 and takes the current release to add to that date. That
way we don't have to maintain anything and as long as we keep a yearly
release we're good. If we change the schedule then we either remove this
or update the logic. That's a risk in shipping code that doesn't fire
for 2 years.

Signed-off-by: Tim Smith <tsmith@chef.io>

### CODE REVIEW

File: lib/chef/client.rb [hunk 1]
```diff
@@ -323,6 +325,22 @@ def run
     # @todo make this stuff protected or private
     #
 
+    # @api private
+    def warn_if_eol
+      require_relative "version"
+
+      # we have a yearly release and we know 15 goes EOL in 2021 so calculate off that
```
Comment (phiggins) [inline lib/chef/client.rb:332]:
```suggestion
      # we have a yearly release and we know 15 goes EOL in 2022 so calculate off that
```

Is this comment wrong?

File: lib/chef/client.rb [hunk 2]
```diff
@@ -323,6 +325,22 @@ def run
...
+      # this way we don't have to update anything or maintain a hash of EOL dates
+      base_eol_year = 2022
+      base_release = Gem::Version.new(15)
+
+      diff_from_base_release = Gem::Version.new(Chef::VERSION).segments.first - base_release.segments.first
```
Comment (lamont-granquist) [inline lib/chef/client.rb:337]: `eol_year = 2007 + Gem::Version.new(Chef::VERSION).segments.first` ?

the double magic constants of 15 expiring in 2022 and doing arithmetic around that seems more confusing to me than just having a magic base year of 2007 that makes the math work out.

File: lib/chef/client.rb (hunk 1 above)
Comment (lamont-granquist) [reply lib/chef/client.rb:332]: isn't that comment right?  15 expires next year.

Comment (phiggins) [reply lib/chef/client.rb:332]: One of the tests is testing that 15 gets an EOL message mentioning 2022. Is the comment right or is the test? Or am I just very confused?

File: lib/chef/client.rb (hunk 2 above)
Comment (lamont-granquist) [reply lib/chef/client.rb:337]: (or 2006)

File: lib/chef/client.rb (hunk 1 above)
Comment (tas50) [reply lib/chef/client.rb]: I set the dates in the specs to be right on the border now so it should line up better with the dates.

File: lib/chef/client.rb (hunk 2 above)
Comment (tas50) [reply lib/chef/client.rb]: Yeah that makes way more sense to me as well. I updated the PR

**Pull request merged**